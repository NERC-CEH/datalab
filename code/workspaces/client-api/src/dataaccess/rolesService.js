import axios from 'axios';
import logger from 'winston';
import config from '../config';
import axiosErrorHandler from '../util/errorHandlers';

const authServiceUrl = config.get('authorisationService');

async function getAllUsersAndRoles(token) {
  logger.debug('Loading all users and roles');

  try {
    const response = await axios.get(`${authServiceUrl}/roles`, { headers: { authorization: token } });

    return response.data.map(roles => ({ ...roles, name: roles.userName }));
  } catch (err) {
    return axiosErrorHandler('Unable to get all users and roles')(err);
  }
}

export default { getAllUsersAndRoles };
