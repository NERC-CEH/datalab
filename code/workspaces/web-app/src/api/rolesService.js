import { gqlQuery, gqlMutation } from './graphqlClient';
import errorHandler from './graphqlErrorHandler';

async function getAllUsersAndRoles() {
  const query = `
    GetAllUsersAndRoles {
      allUsersAndRoles { 
        userId,
        name,
        instanceAdmin,
        dataManager,
        catalogueRole,
        projectRoles {
          projectKey,
          role
        }
      }
    }`;

  const allUsersAndRoles = await gqlQuery(query)
    .then(errorHandler('data.allUsersAndRoles'));
  return allUsersAndRoles;
}

async function setInstanceAdmin(userId, instanceAdmin) {
  const mutation = `
    SetInstanceAdmin($userId: ID!, $instanceAdmin: Boolean!) {
      setInstanceAdmin(userId: $userId, instanceAdmin: $instanceAdmin) {
        userId,
        instanceAdmin
      }
    }`;

  const newInstanceAdmin = await gqlMutation(mutation, { userId, instanceAdmin })
    .then(errorHandler('data.setInstanceAdmin'));
  return newInstanceAdmin;
}

async function setDataManager(userId, dataManager) {
  const mutation = `
    SetDataManager($userId: ID!, $dataManager: Boolean!) {
      setDataManager(userId: $userId, dataManager: $dataManager) {
        userId,
        dataManager
      }
    }`;

  const newDataManager = await gqlMutation(mutation, { userId, dataManager })
    .then(errorHandler('data.setDataManager'));
  return newDataManager;
}

async function setCatalogueRole(userId, catalogueRole) {
  const mutation = `
    SetCatalogueRole($userId: ID!, $catalogueRole: CatalogueRole!) {
      setCatalogueRole(userId: $userId, catalogueRole: $catalogueRole) {
        userId,
        catalogueRole
      }
    }`;

  const newCatalogueRole = await gqlMutation(mutation, { userId, catalogueRole })
    .then(errorHandler('data.setCatalogueRole'));
  return newCatalogueRole;
}

export default { getAllUsersAndRoles, setInstanceAdmin, setDataManager, setCatalogueRole };
