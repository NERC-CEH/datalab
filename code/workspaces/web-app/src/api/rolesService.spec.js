import mockClient from './graphqlClient';
import rolesService from './rolesService';

jest.mock('./graphqlClient');

describe('rolesService', () => {
  beforeEach(() => mockClient.clearResult());

  describe('getAllUsersAndRoles', () => {
    it('should build the correct query and unpack the results', () => {
      mockClient.prepareSuccess({ allUsersAndRoles: 'expectedValue' });

      return rolesService.getAllUsersAndRoles().then((response) => {
        expect(response).toEqual('expectedValue');
        expect(mockClient.lastQuery()).toMatchSnapshot();
      });
    });

    it('should throw an error if the query fails', () => {
      mockClient.prepareFailure('error');

      return rolesService.getAllUsersAndRoles().catch((error) => {
        expect(error).toEqual({ error: 'error' });
      });
    });
  });

  describe('setInstanceAdmin', () => {
    it('should build the correct query and unpack the results', () => {
      mockClient.prepareSuccess({ setInstanceAdmin: 'expectedValue' });

      return rolesService.setInstanceAdmin('one', true).then((response) => {
        expect(response).toEqual('expectedValue');
        expect(mockClient.lastQuery()).toMatchSnapshot();
      });
    });

    it('should throw an error if the query fails', () => {
      mockClient.prepareFailure('error');

      return rolesService.setInstanceAdmin('one', true).catch((error) => {
        expect(error).toEqual({ error: 'error' });
      });
    });
  });

  describe('setDataManager', () => {
    it('should build the correct query and unpack the results', () => {
      mockClient.prepareSuccess({ setDataManager: 'expectedValue' });

      return rolesService.setDataManager('one', true).then((response) => {
        expect(response).toEqual('expectedValue');
        expect(mockClient.lastQuery()).toMatchSnapshot();
      });
    });

    it('should throw an error if the query fails', () => {
      mockClient.prepareFailure('error');

      return rolesService.setDataManager('one', true).catch((error) => {
        expect(error).toEqual({ error: 'error' });
      });
    });
  });

  describe('setCatalogueRole', () => {
    it('should build the correct query and unpack the results', () => {
      mockClient.prepareSuccess({ setCatalogueRole: 'expectedValue' });

      return rolesService.setCatalogueRole('one', 'publisher').then((response) => {
        expect(response).toEqual('expectedValue');
        expect(mockClient.lastQuery()).toMatchSnapshot();
      });
    });

    it('should throw an error if the query fails', () => {
      mockClient.prepareFailure('error');

      return rolesService.setCatalogueRole('one', 'publisher').catch((error) => {
        expect(error).toEqual({ error: 'error' });
      });
    });
  });
});
