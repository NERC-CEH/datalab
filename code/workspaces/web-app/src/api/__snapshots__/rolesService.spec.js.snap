// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rolesService getAllUsersAndRoles should build the correct query and unpack the results 1`] = `
"
    GetAllUsersAndRoles {
      allUsersAndRoles { 
        userId,
        name,
        instanceAdmin,
        dataManager,
        catalogueRole,
        projectRoles {
          projectKey,
          role
        }
      }
    }"
`;

exports[`rolesService setCatalogueRole should build the correct query and unpack the results 1`] = `
"
    SetCatalogueRole($userId: ID!, $catalogueRole: CatalogueRole!) {
      setCatalogueRole(userId: $userId, catalogueRole: $catalogueRole) {
        userId,
        catalogueRole
      }
    }"
`;

exports[`rolesService setDataManager should build the correct query and unpack the results 1`] = `
"
    SetDataManager($userId: ID!, $dataManager: Boolean!) {
      setDataManager(userId: $userId, dataManager: $dataManager) {
        userId,
        dataManager
      }
    }"
`;

exports[`rolesService setInstanceAdmin should build the correct query and unpack the results 1`] = `
"
    SetInstanceAdmin($userId: ID!, $instanceAdmin: Boolean!) {
      setInstanceAdmin(userId: $userId, instanceAdmin: $instanceAdmin) {
        userId,
        instanceAdmin
      }
    }"
`;
